cpp/include/cuml/linear_model/glm.hpp

```diff
@@ -137,85 +141,32 @@ void gemmPredict(const raft::handle_t& handle,
  * @param D                     number of features
  * @param C                     number of outputs (C > 1, for multinomial,
  * indicating number of classes. For logistic and normal, C must be 1.)
- * @param fit_intercept         true if model should include a bias. If true,
- * the initial point/result w0 should point to a memory location of size (D+1) *
- * C
- * @param l1                    l1 regularization strength (if non-zero, will
- * run OWL-QN, else L-BFGS). Note, that as in scikit, the bias will not be
- * regularized.
- * @param l2                    l2 regularization strength. Note, that as in
- * scikit, the bias will not be regularized.
- * @param max_iter              limit on iteration number
- * @param grad_tol              tolerance for gradient norm convergence check.
- * The training process will stop if
- * `norm(current_loss_grad, inf) <= grad_tol * max(current_loss, grad_tol)`.
- * @param change_tol            tolerance for function change convergence check.
- * The training process will stop if
- * `abs(current_loss - previous_loss) <= change_tol * max(current_loss, grad_tol)`,
- * where `previous_loss` is the loss value a small fixed number of steps ago.
- * @param linesearch_max_iter   max number of linesearch iterations per outer
- * iteration
- * @param lbfgs_memory          rank of the lbfgs inverse-Hessian approximation.
- * Method will request memory of size O(lbfgs_memory * D).
- * @param verbosity             verbosity level
  * @param w0                    device pointer of size (D + (fit_intercept ? 1 :
  * 0)) * C with initial point, overwritten by final result.
  * @param f                     host pointer holding the final objective value
- * @param num_iters             host pointer holding the actual number of
- * iterations taken
- * @param loss_type             id of likelihood model (0: logistic/sigmoid, 1:
- * normal/squared, 2: multinomial/softmax)
- * @{
+ * @param num_iters             host pointer holding the actual number of iterations taken
+ * @param sample_weight
  */
+template <typename T>
 void qnFit(const raft::handle_t& cuml_handle,
-           float* X,
-           bool X_col_major,
-           float* y,
-           int N,
-           int D,
-           int C,
-           bool fit_intercept,
-           float l1,
-           float l2,
-           int max_iter,
-           float grad_tol,
-           float change_tol,
-           int linesearch_max_iter,
-           int lbfgs_memory,
-           int verbosity,
-           float* w0,
-           float* f,
-           int* num_iters,
-           int loss_type,
-           float* sample_weight = nullptr);
-void qnFit(const raft::handle_t& cuml_handle,
-           double* X,
+           const qn_params& pams,
+           T* X,
            bool X_col_major,
-           double* y,
+           T* y,
            int N,
```

Comment:
For now, I added the template parameter for indexing only in glm.hpp, so that we don't change the api too often. I think, changing the whole QN is worth a separate PR (maybe we can use mdarray instead of SimpleMat by then?..).